File: internal/pyenv/validate.go
package pyenv

import (
	"bytes"
	"context"
	_ "embed"
	"encoding/json"
	"fmt"
	"os/exec"
	"strings"
)

//go:embed validate.py
var validateScript string

// Verdict is the result of validating one artifact.
type Verdict struct {
	OK     bool     `json:"ok"`
	Stage  string   `json:"stage"`
	Errors []string `json:"errors"`
	// Variables are run()'s parameters, which become invocation variables.
	Variables []Variable `json:"variables"`
}

// Variable is one parameter of run().
type Variable struct {
	Name    string  `json:"name"`
	Type    *string `json:"type"`
	Default *string `json:"default"`
}

// Validate runs the SDK's ScriptValidator against an artifact, with the
// runtime's import list substituted for the SDK's own.
//
// The split is deliberate: the endpoint owns which imports are allowed, and
// the validator owns structure. Trusting the SDK's list would reject
// httpcloak, which ~333 deployed functions import.
func Validate(ctx context.Context, venvDir string, health *Health, source string) (*Verdict, error) {
	if !health.Complete() {
		return nil, fmt.Errorf("cannot validate against a %s runtime report: it carries no import list", health.Status)
	}

	allowed := make([]string, 0, len(health.StdlibModules)+len(health.Packages))
	allowed = append(allowed, health.StdlibModules...)
	for _, p := range health.Packages {
		allowed = append(allowed, p.ImportName)
	}

	request, err := json.Marshal(map[string]any{"source": source, "allowed_imports": allowed})
	if err != nil {
		return nil, err
	}

	cmd := exec.CommandContext(ctx, PythonPath(venvDir), "-c", validateScript)
	cmd.Stdin = bytes.NewReader(request)
	var stdout, stderr bytes.Buffer
	cmd.Stdout, cmd.Stderr = &stdout, &stderr

	if err := cmd.Run(); err != nil {
		return nil, fmt.Errorf("run validator: %w: %s", err, strings.TrimSpace(stderr.String()))
	}

	var v Verdict
	if err := json.Unmarshal(stdout.Bytes(), &v); err != nil {
		return nil, fmt.Errorf("decode validator output: %w (stdout: %q)", err, stdout.String())
	}
	return &v, nil
}

File: internal/pyenv/validate.py
"""Validate a bundled artifact against the runtime's own rules.

Reads {"source": str, "allowed_imports": [str]} on stdin and writes a verdict
as JSON on stdout. Never raises for a rejected script: a non-zero exit would
be indistinguishable from the interpreter or the SDK being broken.

Why the allow list is injected rather than trusted:

The published notte-sdk carries its own ALLOWED_IMPORTS, and it is stale. At
1.8.31 it is an explicit 41-entry list with no httpcloak, httpx, bs4 or tqdm,
and it *includes* tempfile, which the runner discards. Running it unmodified
rejects functions that are deployed and serving traffic. So the validator's
machinery is reused for what it gets right - the entry point, forbidden nodes,
forbidden calls, relative imports - while the import question is answered from
GET /functions/health, which asks the runner.
"""

import ast
import json
import sys


def main() -> None:
    request = json.load(sys.stdin)
    source = request["source"]

    try:
        from notte_core.ast import ScriptValidator
    except Exception as exc:  # pragma: no cover - environment, not input
        json.dump({"ok": False, "stage": "import", "errors": [f"cannot import notte_core.ast: {exc}"]}, sys.stdout)
        return

    # The runtime's list replaces the SDK's, and the SDK's denylist is emptied
    # so it cannot veto a name the runtime allows. Anything not in the runtime
    # list is simply absent, and check_valid_import rejects it.
    ScriptValidator.ALLOWED_IMPORTS = set(request["allowed_imports"])
    if hasattr(ScriptValidator, "DISALLOWED_STDLIB_IMPORTS"):
        ScriptValidator.DISALLOWED_STDLIB_IMPORTS = set()

    errors = []

    # parse_script accepts two top-level run() definitions where the server
    # rejects them, so that is checked here rather than delegated.
    try:
        tree = ast.parse(source)
        runs = [n for n in tree.body if isinstance(n, (ast.FunctionDef, ast.AsyncFunctionDef)) and n.name == "run"]
        if len(runs) > 1:
            lines = ", ".join(str(n.lineno) for n in runs)
            errors.append(f"multiple top-level run() definitions (lines {lines}); the runtime accepts exactly one")
    except SyntaxError as exc:
        json.dump({"ok": False, "stage": "syntax", "errors": [f"line {exc.lineno}: {exc.msg}"]}, sys.stdout)
        return

    variables = []
    try:
        info = ScriptValidator.parse_script(source, restricted=True)
        variables = [
            {"name": v.name, "type": v.type, "default": v.default}
            for v in info.variables
        ]
    except Exception as exc:
        errors.append(f"{type(exc).__name__}: {exc}")

    json.dump({"ok": not errors, "stage": "validate", "errors": errors, "variables": variables}, sys.stdout)


main()
